src/MIDebugEngine/AD7.Impl/AD7Thread.cs

```diff
@@ -282,23 +282,27 @@ int IDebugThread2.Resume(out uint suspendCount)
         }
 
         // Sets the next statement to the given stack frame and code context.
-        // https://docs.microsoft.com/en-us/visualstudio/extensibility/debugger/reference/idebugthread2-setnextstatement
         int IDebugThread2.SetNextStatement(IDebugStackFrame2 stackFrame, IDebugCodeContext2 codeContext)
         {
-            // VS does provide a frame so at least do some sanity checks
-            AD7StackFrame frame = stackFrame as AD7StackFrame;
-            if (frame != null && (frame.ThreadContext.Level != 0 || frame.Thread != this))
-                return Constants.S_FALSE;
-
-            try
+            ulong addr = ((AD7MemoryAddress)codeContext).Address;
+            AD7StackFrame frame = ((AD7StackFrame)stackFrame);
+            if (frame.ThreadContext.Level != 0 || frame.Thread != this || !frame.ThreadContext.pc.HasValue)
             {
-                ulong addr = ((AD7MemoryAddress)codeContext).Address;
-                return _engine.Jump(addr);
+                return Constants.S_FALSE;
             }
-            catch (Exception)
+            string toFunc = EngineUtils.GetAddressDescription(_engine.DebuggedProcess, addr);
+            string fromFunc = EngineUtils.GetAddressDescription(_engine.DebuggedProcess, frame.ThreadContext.pc.Value);
+            if (toFunc != fromFunc)
             {
                 return Constants.S_FALSE;
             }
+            string result = frame.EvaluateExpression("$pc=" + EngineUtils.AsAddr(addr, _engine.DebuggedProcess.Is64BitArch));
```

Comment:
This function is now only used by VS, right?
It can handle this expression?

Reply:
> This function is now only used by VS, right?

Correct. Currently this code isn't reachable in DAP scenarios.

> It can handle this expression?

This is just an internal call within MIEngine, so it will not matter what IDE is used. It worked with whatever scenario Set Next Statement was originally tested (Android?). I don't know if it has been tested more broadly. Certainly using the actual GDB command is better. But this PR just seeks to finish off disassembly support. Actually implementing set next statement is still left for your PR. Someone needs to finish off handling the work of handling that stopping event...

Reply:
Ah yeah my bad, of course this actually goes down to gdb and not up. Setting `$pc` is actually different from the jump command in that it `does not start your program running; it only changes the address of where it will run when you continue`: https://sourceware.org/gdb/current/onlinedocs/gdb/Jumping.html
So I'm not sure how this worked out regarding UI/UX.

Reply:
Yup, we would need to hide this difference -- the new set next statement would need to save some state in the process object to indicate that it is waiting for that stop event, and then SetNextStatement will need to wait for the stop event to be raised (or for the target process to exit).